services/src/main/java/org/keycloak/services/clientpolicy/DefaultClientPolicyManager.java

```diff
@@ -0,0 +1,157 @@
+        for (ClientPolicyProvider policy : getProviders(realm)) {
+            ClientPolicyLogger.logv(logger, "Policy Operation : name = {0}, provider id = {1}", policy.getName(), policy.getProviderId());
+            if (!isSatisfied(policy, condition)) continue;
+            execute(policy, executor);
+        }
+    }
+
+    private List<ClientPolicyProvider> getProviders(RealmModel realm) {
+        List<ClientPolicyProvider> providers = providersMap.get(realm.getId());
+        if (providers == null) {
+            providers = new LinkedList<>();
+            List<ComponentModel> policyModels = realm.getComponents(realm.getId(), ClientPolicyProvider.class.getName());
+            for (ComponentModel policyModel : policyModels) {
+                try {
+                    ClientPolicyProvider policy = session.getProvider(ClientPolicyProvider.class, policyModel);
+                    ClientPolicyLogger.logv(logger, "Loaded Policy Name = {0}", policyModel.getName());
+                    session.enlistForClose(policy);
+                    providers.add(policy);
+                } catch (Throwable t) {
+                    logger.errorv(t, "Failed to load provider {0}", policyModel.getId());
+                }
+            }
+            providersMap.put(realm.getId(), providers);
+        } else {
+            ClientPolicyLogger.log(logger, "Use cached policies.");
+        }
+        return providers;
+    }
+
+    private boolean isSatisfied(
+            ClientPolicyProvider policy,
+            ClientConditionOperation op) {
+
+        List<ClientPolicyConditionProvider> conditions = policy.getConditions();
+
+        if (conditions == null || conditions.isEmpty()) {
+            ClientPolicyLogger.log(logger, "NEGATIVE :: This policy is not applied. No condition exists.");
+            return false;
+        }
+
+        boolean ret = false;
+        for (ClientPolicyConditionProvider condition : conditions) {
+            try {
+                if (!op.run(condition)) {
+                    ClientPolicyLogger.logv(logger, "NEGATIVE :: This policy is not applied. condition not satisfied. name = {0}, provider id = {1}, ", condition.getName(), condition.getProviderId());
+                    return false;
+                } else {
+                    ret = true;
+                }
+            } catch (ClientPolicyException cpe) {
+                if (cpe.getError().equals(ClientPolicyConditionProvider.SKIP_EVALUATION)) {
+                    ClientPolicyLogger.logv(logger, "SKIP : This condition is not evaluated due to its nature. name = {0}, provider id = {1}", condition.getName(), condition.getProviderId());
+                    continue;
+                }
+                ClientPolicyLogger.logv(logger, "CONDITION EXCEPTION : name = {0}, provider id = {1}, error = {2}, error_detail = {3}", condition.getName(), condition.getProviderId(), cpe.getError(), cpe.getErrorDetail());
+                return false;
```

Comment:
Thinking about this some more I wonder if skipping the policies on an exception is a good thing. It could result in some error in a condition resulting in accepting things that shouldn't be accepted. Wonder if we should fail the request in this case instead.

Reply:
Yes, I'll revise codes to make the request fail if `ClientPolicyException`is thrown from `ClientPolicyConditionProvider`.